core-processor/src/main/java/io/micronaut/expressions/parser/ast/access/CandidateMethod.java

```diff
@@ -0,0 +1,190 @@
+    }
+
+    public boolean isMatching(VisitorContext ctx) {
+        int totalParams = parameterTypes.size();
+        int totalArguments = argumentTypes.size();
+
+        if (totalParams == 0) {
+            return totalArguments == 0;
+        } else if (totalArguments < totalParams - 1) {
+            return false;
+        } else {
+            ClassElement lastParameter = getLastParameter();
+            boolean varargsCandidate = methodElement.isVarArgs() ||
+                                           (lastParameter.isArray() && lastParameter.getArrayDimensions() == 1);
+
+            if (varargsCandidate) {
+                this.varargsIndex = calculateVarargsIndex(ctx);
+                return varargsIndex != -1;
+            }
+
+            if (totalArguments != totalParams) {
+                return false;
+            }
+
+            for (int i = 0; i < parameterTypes.size(); i++) {
+                ClassElement argumentType = argumentTypes.get(i);
+                ClassElement parameterType = parameterTypes.get(i);
+
+                if (!isAssignable(parameterType, argumentType)) {
+                    return false;
+                }
+            }
+            return true;
+        }
+    }
+
+    private int calculateVarargsIndex(VisitorContext ctx) {
+        ClassElement lastParameter = CollectionUtils.last(parameterTypes);
+        if (lastParameter == null) {
+            return -1;
+        }
+
+        for (int i = 0; i < parameterTypes.size(); i++) {
+            ClassElement parameterType = parameterTypes.get(i);
+            boolean isLastParameter = i == parameterTypes.size() - 1;
+
+            if (!isLastParameter) {
+                if (argumentTypes.size() < i) {
+                    return -1;
+                }
+
+                ClassElement argumentType = argumentTypes.get(i);
+                if (!isAssignable(parameterType, argumentType)) {
+                    return -1;
+                }
+            } else {
+                parameterType =
+                    getRequiredClassElement(getTypeReference(parameterType).getElementType(), ctx);
+
+                if (argumentTypes.size() < i) {
+                    return argumentTypes.size();
+                }
+
+                for (int k = i; k < argumentTypes.size(); k++) {
+                    ClassElement argumentType = argumentTypes.get(i);
+                    if (!isAssignable(parameterType, argumentType)) {
+                        return -1;
+                    }
+                }
+
+                return i;
+            }
+        }
+
+        return -1;
+    }
+
+    public Method toAsmMethod() {
+        String description =
+            methodElement.getReturnType().getDescription(false) + " "
+                + methodElement.getSimpleName()
+                + Arrays.stream(methodElement.getParameters())
+                      .map(parameter -> {
+                          String name = parameter.getType().getName();
```

Comment:
can you re-use https://github.com/micronaut-projects/micronaut-core/blob/f7a46b8996342cd5de1552cb9d939ab64058945c/core-processor/src/main/java/io/micronaut/inject/writer/AbstractClassFileWriter.java#L1188

This custom logic is likely to result in correct descriptors for different array dimensions